File: backend/app/routers/fixture_import.py
from fastapi import APIRouter, Depends, File, UploadFile
from fastapi.responses import StreamingResponse
from io import BytesIO

from app.core.deps import DbSession, require_admin
from app.models.user import User
from app.schemas.fixture_import import (
    FixtureImportCommitResponse,
    FixtureImportPreviewResponse,
)
from app.services.fixture_import_service import (
    commit_fixture_import,
    create_fixture_template_excel,
    preview_fixture_import,
)


router = APIRouter(prefix="/fixture-import", tags=["fixture-import"])


@router.get("/template")
def download_fixture_template(
    _: User = Depends(require_admin),
):
    content = create_fixture_template_excel()

    headers = {
        "Content-Disposition": 'attachment; filename="plantilla_fixture_prode.xlsx"'
    }

    return StreamingResponse(
        BytesIO(content),
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers=headers,
    )


@router.post("/preview", response_model=FixtureImportPreviewResponse)
async def preview_import(
    file: UploadFile = File(...),
    _: User = Depends(require_admin),
):
    return await preview_fixture_import(file)


@router.post("/commit", response_model=FixtureImportCommitResponse)
async def commit_import(
    db: DbSession,
    file: UploadFile = File(...),
    _: User = Depends(require_admin),
):
    return await commit_fixture_import(db, file)

File: frontend/src/api/fixtureImport.ts
import { api } from "./client";
import type {
  FixtureImportCommitResponse,
  FixtureImportPreviewResponse,
} from "../types/fixtureImport";

export async function downloadFixtureTemplate(): Promise<void> {
  const response = await api.get("/fixture-import/template", {
    responseType: "blob",
  });

  const blob = new Blob([response.data], {
    type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  });

  const url = window.URL.createObjectURL(blob);

  const link = document.createElement("a");
  link.href = url;
  link.download = "plantilla_fixture_prode.xlsx";
  document.body.appendChild(link);
  link.click();
  link.remove();

  window.URL.revokeObjectURL(url);
}

export async function previewFixtureImport(
  file: File,
): Promise<FixtureImportPreviewResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await api.post<FixtureImportPreviewResponse>(
    "/fixture-import/preview",
    formData,
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    },
  );

  return response.data;
}

export async function commitFixtureImport(
  file: File,
): Promise<FixtureImportCommitResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await api.post<FixtureImportCommitResponse>(
    "/fixture-import/commit",
    formData,
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    },
  );

  return response.data;
}